### DISCUSSION TITLE
Should I use Guava or Mug or Both?

### DISCUSSION THREAD

**Discussion by xingyutangyuan** [google/mug#59] [General]
Guava and Mug are both general-purpose library. And there is also guava-mug artifact.

What are some guidelines on using either one and do I add both as my dependencies? Is there overlaps?

**Comment by fluentfuture** [comment]
That's a great question.

Within Google we use Guava like breathing air. `ImmutableList`, `ImmutableMap` are everywhere. I also frequently use `Preconditions`, `CharMatcher`, and some useful packages: `com.google.common.primitives`,`com.google.common.io`. And while not used very often, `com.google.common.hash` is great when I need to do hashing.

Now assuming I'm writing code outside of Google, chances are Guava is already in the dependencies anyways. But if it's not, I might need to be careful because some version of it may already exist in the transitive dependency closure and I don't want version conflicts.

Mug on the other hand is less versatile. But for what it does well, its usage has been ramping up rapidly inside Google. This includes:

* `BiStream` + `BiCollectors` - almost anyone who uses `Map` can take advantage of BiStream (if they like Java streams)
* `Substring` - Guava hasn't focused much on adding rich string manipulation utils (internally there used to be a  _StringUtils_ class with like 76 random static util methods. It was deprecated as such classes are easy to become a dumping ground)
* `StringFormat` - It's both a convenient parser and a template. Guava historically decided not to add string templating. But with lambda and ErrorProne compile-time checks, it turns out pleasantly useful.

The recent addition of `DateTimeFormats`, which you helped in the making, is also a nice utility I'd personally recommend. It saves time learning the Java date time formatting syntax and makes the time format WYSIWYG.

**Comment by xingyutangyuan** [comment]
Thanks for the guideline!

Any thoughts between `Splitter` and the split mehtods of `Substring`? Seems like both do string splitting.

**Reply by fluentfuture** [reply]
Oh yeah `Substring` does do lots of what Splitter does. 

The bottom line: I wouldn't pull in extra Mug dependency if Splitter already works, and vice versa.

When I have both deps,I tend to use `Substring` (`StringFormat` if it's based on a compile-time format string) for almost all splitting work. I guess because `Substring` does more than just splitting, and it's natural to use the same abstraction for all of finding, matching, removing, splitting and replacing without switching gear.
